applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp

```diff
@@ -83,183 +249,27 @@ void DivDGMaxEigenvalue<DIM>::solve(
-        builder.setMatrixExtractor([&](const Base::Face* face) {
-            const Base::FaceMatrix& faceMatrix = face->getFaceMatrix(
-                DivDGMaxDiscretization<DIM>::FACE_STIFFNESS_MATRIX_ID);
-            LinearAlgebra::MiddleSizeMatrix block1, block2;
-            block1 = faceMatrix.getElementMatrix(Base::Side::LEFT,
-                                                 Base::Side::RIGHT);
-            block2 = faceMatrix.getElementMatrix(Base::Side::RIGHT,
-                                                 Base::Side::LEFT);
-
-            return std::make_pair(block1, block2);
-        });
-        kphaseshifts = builder.build(indexing);
-        DGMaxLogger(VERBOSE, "Initialized boundary shifting");
-    }
-
-    // Setting up eigenvalue solver //
-    //////////////////////////////////
-
-    DGMaxLogger(INFO, "Configuring Eigenvalue solver");
-    EPS eigenSolver;
-    error = EPSCreate(PETSC_COMM_WORLD, &eigenSolver);
-    CHKERRABORT(PETSC_COMM_WORLD, error);
-
-    // SH 180221
-    error = EPSSetOperators(eigenSolver, massMatrix, stiffnessMatrix);
-    // MatSetOption(product, MAT_NEW_NONZERO_LOCATIONS, PETSC_TRUE);
-    // MatSetOption(stiffnessMatrix, MAT_NEW_NONZERO_LOCATIONS, PETSC_TRUE);
-    CHKERRABORT(PETSC_COMM_WORLD, error);
-    // SH 180212
-    // error = EPSSetUp(eigenSolver_);
-    CHKERRABORT(PETSC_COMM_WORLD, error);
-
-    // As the stiffness and mass matrix are switched (i.e. M u = lambda S u),
-    // the eigenvalues correspond to omega^{-2}, thus the lowest bands give the
-    // largest eigenvalues.
-    EPSSetWhichEigenpairs(eigenSolver, EPS_LARGEST_MAGNITUDE);
-
-    EPSSetProblemType(eigenSolver, EPS_GNHEP);
-    DGMaxLogger(INFO, "Eigenvalue solver configured");
-
-    // Storage vectors //
-    /////////////////////
-
-    Vec* eigenVectors;
-    eigenVectors =
-        new Vec[numberOfEigenVectors];  // a few extra in case SLEPc finds more
-                                        // than the requested amount of
-                                        // eigenvalues
-    error = VecDuplicateVecs(globalVector, numberOfEigenVectors, &eigenVectors);
-    CHKERRABORT(PETSC_COMM_WORLD, error);
-    int converged = 0;
-    DGMaxLogger(INFO, "Storage vectors assembled");
+    Workspace<DIM> workspace(&mesh_);
 
     std::size_t outputId = 0;
     std::size_t expectedNumberOfSteps = driver.getNumberOfKPoints();
-    // For testing
-    //    maxSteps = 21;
-
-    std::vector<PetscScalar> eigenvalues(numberOfEigenvalues);
-    DivDGMaxResult<DIM> result;
 
     DGMaxLogger(INFO, "Starting k-vector walk");
     for (std::size_t solve = 0; !driver.stop(); driver.nextKPoint(), ++solve) {
         LinearAlgebra::SmallVector<DIM> currentK = driver.getCurrentKPoint();
         numberOfEigenvalues = driver.getTargetNumberOfEigenvalues();
         DGMaxLogger(INFO, "Solving for k-vector %/%", solve + 1,
                     expectedNumberOfSteps);
+        workspace.solve(currentK, numberOfEigenvalues);
```

Comment:
Yeah it is a bit counterintuitive for me or I do not get the whole picture. Why doesn't the solver `solve` instead of the `workspace`?

Reply:
There are multiple differences. The overall solve (many k-points) has multiple steps:
 - Global setup (e.g. create matrices)
 - For each k-point
   - Setup for that point
   - Solve for that point
   - Output the results
 - Cleanup
 
All most all of the individual steps are done by separate methods in the Workspace. What you are seeing here is the solve at a single k-point.